interfaces/machine_parameters/ring.py

```diff
@@ -0,0 +1,588 @@
+                self.alpha_1[i] - self.alpha_0[i] * self.eta_0[i]
+
+    def _eta2(self):
+        """ Function to calculate the second order slippage factor eta_2 """
+
+        for i in range(0, self.n_sections):
+            self.eta_2[i] = - self.beta[i]**2 * (5 * self.beta[i]**2 - 1) / \
+                (2 * self.gamma[i]**2) + self.alpha_2[i] - 2 * self.alpha_0[i] *\
+                self.alpha_1[i] + self.alpha_1[i] / self.gamma[i]**2 + \
+                self.alpha_0[i]**2 * self.eta_0[i] - 3 * self.beta[i]**2 * \
+                self.alpha_0[i] / (2 * self.gamma[i]**2)
+
+    def parameters_at_time(self, cycle_moments):
+        """ Function to return various cycle parameters at a specific moment in
+        time. The cycle time is defined to start at zero in turn zero.
+
+        Parameters
+        ----------
+        cycle_moments : float array
+            Moments of time at which cycle parameters are to be calculated [s].
+
+        Returns
+        -------
+        parameters : dictionary
+            Contains 'momentum', 'beta', 'gamma', 'energy', 'kin_energy',
+            'f_rev', 't_rev'. 'omega_rev', 'eta_0', and 'delta_E' interpolated
+            to the moments contained in the 'cycle_moments' array
+
+        """
+
+        parameters = {}
+        parameters['momentum'] = np.interp(cycle_moments, self.cycle_time,
+                                           self.momentum[0])
+        parameters['beta'] = np.interp(cycle_moments, self.cycle_time,
+                                       self.beta[0])
+        parameters['gamma'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.gamma[0])
+        parameters['energy'] = np.interp(cycle_moments, self.cycle_time,
+                                         self.energy[0])
+        parameters['kin_energy'] = np.interp(cycle_moments, self.cycle_time,
+                                             self.kin_energy[0])
+        parameters['f_rev'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.f_rev)
+        parameters['t_rev'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.t_rev)
+        parameters['omega_rev'] = np.interp(cycle_moments, self.cycle_time,
+                                            self.omega_rev)
+        parameters['eta_0'] = np.interp(cycle_moments, self.cycle_time,
+                                        self.eta_0[0])
+        if len(self.cycle_time) == len(self.delta_E[0]):
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time,
+                                              self.delta_E[0])
+        else:
+            parameters['delta_E'] = np.interp(cycle_moments,
+                                              self.cycle_time[:-1],
+                                              self.delta_E[0])
```

Comment:
I really don't like this, but couldn't think of a better solution.  If you can think of one I'll be grateful...

Reply:
I think this came from the fact that the definition of n_turns was different if the data was passed by_turn or by_time in preprocess, now that all programs should all be in size n_turns+1 except delta_E which is of size n_turns, we can keep only the 'else' statement

Reply:
What about having kwargs to select which attributes to get from this function?

Reply:
No it's the length thing with `n_turns` or `n_turns+1`.  I did it, I just don't like it!